internal/sopsenv/sopsenv.go

```diff
@@ -0,0 +1,283 @@
+	"sort"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"github.com/giantswarm/konfigure/internal/sopsenv/key"
+
+	// GS stuff uses `kgs`-generated kubeconfigs that use
+	// `oidc` auth provider. This import makes is possible to
+	// run `konfigure` locally for troubleshooting purposes.
+	_ "k8s.io/client-go/plugin/pkg/client/auth/oidc"
+)
+
+const (
+	// Environment variables used to configure SOPS, and to point it to the
+	// keys storage. This helps creating temporary and "isolated" environment
+	// for SOPS, in a similar manner how Flux does it. This way we make sure
+	// we do not interfere with Kustomize operations, by leaving key chains
+	// under default locations expected by SOPS, nor by leaving env vars set
+	// to some custom locations when Kustomize decrypts files.
+	ageKeyFileVar = "SOPS_AGE_KEY_FILE"
+	gnuPGHomeVar  = "GNUPGHOME"
+
+	// `konfigure.giantswarm.io/data=sops-key` is used to fetch Kubernetes
+	// Secrets with SOPS keys in order to import them to a temporary location.
+	konfigureLabelKey   = "konfigure.giantswarm.io/data"
+	konfigureLabelValue = "sops-keys"
+	konfigureNamespace  = "giantswarm"
+	konfigureTmpDirName = "konfigure-sops-"
+
+	// Keys extensions supported
+	secretPGPExt = ".asc"
+	secretAgeExt = ".agekey"
+)
+
+type SOPSEnvConfig struct {
+	K8sClient  kubernetes.Interface
+	KeysDir    string
+	KeysSource string
+	Logger     micrologger.Logger
+}
+
+type SOPSEnv struct {
+	cleanup    func()
+	k8sClient  kubernetes.Interface
+	keysDir    string
+	keysSource string
+	logger     micrologger.Logger
+}
+
+// NewSOPSEnv creates SOPS environment configurator, it works according to the
+// below combinations.
+//
+// 1. User expects to run SOPS against his local default keychains for GPG and AGE:
+//      keysDir=""
+//	    keysSource="local"
+//
+// 2. User expects to run SOPS against his custom keychains located under `path`:
+//      keysDir="path"
+//      keysSource="local"
+//
+// 3. User expects to run SOPS against  Kubernetes-downloaded keys stored at tmp location:
+//      keysDir=""
+//      keysSource="kubernetes"
+//
+// 4. User expects to run SOPS against Kubernetes-downloaded keys stored under `path`:
+//      keysDir="path"
+//      keysSource="kubernetes"
+//
+func NewSOPSEnv(config SOPSEnvConfig) (*SOPSEnv, error) {
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	s := &SOPSEnv{
+		keysDir:    config.KeysDir,
+		keysSource: config.KeysSource,
+		logger:     config.Logger,
+	}
+
+	if config.KeysSource == key.KeysSourceLocal {
+		return s, nil
+	}
+
+	if config.KeysDir == "" {
+		keysDir, err := os.MkdirTemp("", konfigureTmpDirName)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+
+		s.keysDir = keysDir
+		s.cleanup = func() { os.RemoveAll(keysDir) }
```

Comment:
Should we have a clean up for the fetched data when the folder is given by the user? I guess it could get complicated as you only want to clean up the konfigure fetched files and not clean up the whole directory recursively in that case.

Reply:
I don't believe we should clean user-given directory as it's beyond what `SOPSEnv` suppose to do, I think. First, it fits some use cases, like troubleshooting, or running inside the cluster where we pre-fetch keys to a given location, but of course it is kind of a trade off, because ideally it would be to not have keys stored on users laptops. That why we will rather recommend not using `sops-keys-dir` flag until needed by the above cases, and by a "conscious" user knowning what he does.